File: 2020/dest.py
import math


class Node:
    def __init__(self, zatopeno=False):
        self.visited = False
        self.distance = math.inf

        self.connected_nodes = []
        self.zatopeno = zatopeno

    def connect(self, other):
        self.connected_nodes.append(other)
        other.connected_nodes.append(self)


def init(priklad_choose=1):

    if priklad_choose == 1:
        # S inputem

        nodes_count, paths_count, limit_count = [
            int(x) for x in input().split(" ")]

        zatopene_nodes = [int(x) for x in input().split(" ")]

        nodes = [Node() if zatopene_nodes[i] == 0 else Node(True)
                 for i in range(nodes_count)]

        for i in range(paths_count):
            first, second = [int(x) for x in input().split(" ")]
            nodes[first].connect(nodes[second])

    elif priklad_choose == 2:
        # Easy příklad impossible

        nodes_count, paths_count, limit_count = 5, 4, 2
        zatopene_nodes = [int(x) for x in "0 1 0 0 0".split(" ")]
        join_nodes = ["0 1", "1 2", "2 3", "3 4"]

    elif priklad_choose == 3:
        # Easy příklad possible

        nodes_count, paths_count, limit_count = 5, 4, 2
        zatopene_nodes = [int(x) for x in "0 0 0 1 1".split(" ")]
        join_nodes = ["0 1", "1 2", "2 3", "3 4"]

    elif priklad_choose == 4:
        # Harder z papieru

        nodes_count, paths_count, limit_count = 10, 14, 2
        zatopene_nodes = [int(x) for x in "0 0 1 1 0 0 1 0 0 0".split(" ")]
        join_nodes = ["0 1", "0 2", "0 3", "1 4", "1 6", "2 5", "3 5",
                      "3 7", "4 5", "5 6", "5 7", "6 8", "7 9", "8 9"]
    elif priklad_choose == 5:
        # Problematikos

        nodes_count, paths_count, limit_count = 11, 11, 2
        zatopene_nodes = [int(x) for x in "0 0 0 0 0 0 1 0 0 0 0".split(" ")]
        join_nodes = ["0 1", "0 6", "1 2", "2 3", "3 4", "4 5", "5 8",
                      "6 7", "7 8", "8 9", "9 10"]

    elif priklad_choose == 6:
        # Problematikos grandos

        nodes_count, paths_count, limit_count = 24, 36, 6
        zatopene_nodes = [
            int(x) for x in "0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0".split(" ")]
        join_nodes = ["0 1", "0 3", "0 6", "1 2", "1 3", "1 4", "2 6", "2 7", "3 8", "4 7",  # 10
                      "5 8", "6 9", "6 10", "7 11", "7 12", "8 13", "8 14", "9 15", "10 15",  # 9
                      "10 16", "11 15", "11 17", "12 17", "12 18", "13 18", "14 18", "15 20",  # 8
                      "15 21", "16 19", "17 20", "18 21", "18 22", "19 23", "20 23", "21 23", "22 23"]  # 9

    if priklad_choose != 1:
        nodes = [Node() if zatopene_nodes[i] == 0 else Node(True)
                 for i in range(nodes_count)]

        for join in join_nodes:
            first, second = join.split(" ")
            first, second = int(first), int(second)

            nodes[first].connect(nodes[second])

    nodes[-1].distance = 0
    queue = [nodes[-1]]
    min_dist = -1

    while len(queue) != 0:
        curr = queue.pop(0)

        for next in curr.connected_nodes:
            if not next.visited:
                if not(curr.zatopeno and curr.distance > limit_count):
                    queue.append(next)

                    next.distance = curr.distance + 1
                    next.visited = True

        if not(curr.zatopeno and curr.distance > limit_count):
            if curr == nodes[0]:
                min_dist = curr.distance
                break

    print(min_dist)


init(4)

File: 2020/dest.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <string>
#include <sstream>
#include <bits/stdc++.h>

using namespace std;

class Node
{
public:
    bool visited = false;
    int distance = numeric_limits<int>::max();
    vector<Node *> connected_nodes;
    bool zatopeno;

    Node(bool aZatopeno)
    {
        zatopeno = aZatopeno;
    }
};

void init()
{
    int nodes_count, paths_count, limit_count;

    // Nodes, Paths, Limit input
    cin >> nodes_count >> paths_count >> limit_count;

    // Zatopene Nodes
    int zatopene_nodes[nodes_count];
    string zatopenost_str;

    cin.ignore();
    getline(cin, zatopenost_str);

    string buf;                      // Have a buffer string
    stringstream ss(zatopenost_str); // Insert the string into a stream

    vector<string> zatopeno; // Create vector to hold our words

    while (ss >> buf)
        zatopeno.push_back(buf);

    // Ukládám: stoi() -> str to int

    // Nodes array
    vector<Node *> nodes;

    for (int i = 0; i < nodes_count; i++)
    {
        if (zatopeno[i] == "1")
        {
            nodes.push_back(new Node(true));
        }
        else
        {
            nodes.push_back(new Node(false));
        }
    }

    // Paths connection
    int paths[paths_count * 2];
    int first_input, second_input;

    for (int i = 0; i < paths_count; i++)
    {
        cin >> first_input >> second_input;
        paths[i * 2] = first_input;
        paths[i * 2 + 1] = second_input;
    }

    for (int i = 0; i < paths_count; i += 1)
    {
        Node *node1 = nodes[paths[i * 2]];
        Node *node2 = nodes[paths[i * 2 + 1]];
        node1->connected_nodes.push_back(node2);
        node2->connected_nodes.push_back(node1);
        // nodes[paths[i + 1]]->connect(nodes[paths[i]]);
    }

    // Testing
    /*
    cout << "Nodes: " << nodes_count << " Paths: " << paths_count << " Limit: " << limit_count << endl;
    cout << "Zatopenost: " << endl;
    for (int i = 0; i < nodes_count; i++)
    {
        cout << zatopeno[i] << endl;
    }

    for (int i = 0; i < paths_count; i++)
    {
        cout << paths[i * 2] << "<->" << paths[i * 2 + 1] << endl;
    }

    for (int i = 0; i < nodes_count; i++)
    {
        cout << nodes[i] << endl;
    }

    for (int i = 0; i < nodes_count; i++)
    {
        cout << nodes[i]->zatopeno << endl;

        cout << nodes[i]->connected_nodes[0] << endl;
        if (nodes[i]->connected_nodes.size() > 1)
        {
            cout << nodes[i]->connected_nodes[1] << endl;
        }
        if (nodes[i]->connected_nodes.size() > 2)
        {
            cout << nodes[i]->connected_nodes[2] << endl;
        }
    }*/

    nodes.back()->distance = 0;
    vector<Node *> queue;
    queue.push_back(nodes.back());

    // Experimental
    for (int i = 0; i < nodes.back()->connected_nodes.size(); i++)
    {
        Node *next = nodes.back()->connected_nodes[i];
        auto finder = find(next->connected_nodes.begin(), next->connected_nodes.end(), nodes.back());
        int index = distance(next->connected_nodes.begin(), finder);

        next->connected_nodes.erase(next->connected_nodes.begin() + index);
    }

    int min_distance = -1;
    int steps = 0;

    while (queue.size() != 0)
    {
        Node *curr = queue.front();
        queue.erase(queue.begin());

        for (int i = 0; i < curr->connected_nodes.size(); i++)
        {
            steps++;
            Node *next = curr->connected_nodes[i];
            if (!next->visited)
            {
                if (!(curr->zatopeno && curr->distance > limit_count))
                {
                    queue.push_back(next);

                    next->distance = curr->distance + 1;
                    next->visited = true;
                }
            }
        }
        if (!(curr->zatopeno && curr->distance > limit_count))
        {
            if (curr == nodes[0])
            {
                min_distance = curr->distance;
                break;
            }
        }
    }
    cout << min_distance << endl;
    cout << "steps: " << steps << endl;
};

int main()
{
    init();
}